nio-impl/src/main/java/org/xnio/nio/WorkerThread.java

```diff
@@ -757,7 +780,7 @@ void cancelKey(final SelectionKey key) {
         }
     }
 
-    void setOps(final SelectionKey key, final int ops) {
+    synchronized void setOps(final SelectionKey key, final int ops) {
```

Comment:
I'm not sure what synchronizing on "this" does here.  Which data is being protected by this synchronization?

Reply:
The set of interestOps: The call to key.interestOps(key.interestOps() | ops);

Reply:
I see.  Synchronizing on the thread to modify the interest ops may be too coarse though; at the very least I would call it "weird".  I think it would be better to synchronize on the SelectionKey itself, and only for the duration of the read-modify-write operation (not during the calls to selector or queueTask()).  WDYT?